### PR TITLE
Add protocol and timeout argument on DNSClient connection

### PR DESCRIPTION
**Pull request by jnahelou**
Hello,

Sometime DNS queries failed with following error (due to latency between master and resolvers)
 ```
* dns_cname_record.cname: dns_cname_record.cname: Error querying DNS record: dial tcp xx.xx.xx.xx:53: i/o timeout                                                                               
```

Following [github.com/miekg/dns/](https://github.com/miekg/dns/), default timeout is set to 2s.
Using tc, it's easy to verify it : 
```
$ sudo tc qdisc add dev wlp1s0 root netem delay 3s 
$ terraform apply 
dns_cname_record.cname: Refreshing state... (ID: foo.localdomain.)                 
Error refreshing state: 1 error(s) occurred:    

* dns_cname_record.cname: 1 error(s) occurred:  

* dns_cname_record.cname: dns_cname_record.cname: Error querying DNS record: dial tcp xx.xx.xx.xx:53: i/o timeout    
```

This pull request add 2  optional keys : 
* protocol : By default udp protocol is set, this property allow user to change to : 

> https://golang.org/pkg/net/  
Allowed values are"tcp", "tcp4" (IPv4-only), "tcp6" (IPv6-only), "udp", "udp4" (IPv4-only), "udp6" (IPv6-only), "ip", "ip4" (IPv4-only), "ip6" (IPv6-only), "unix", "unixgram" and "unixpacket".

Variable name is ```protocol``` because ```net``` is not explicit.

* timeout : change default timeout (second)

Usage : 
```main.tf
provider "dns" {
  update {
    server = "xx.xx.xx.xx"
    protocol = "tcp"
    timeout = 5
  }
}
```

### CODE REVIEW

**Comment by jnahelou** [comment]
Up

**Comment by jnahelou** [comment]
Thank's for this feedback, I updated my version following #35 fix.

File: dns/config.go [hunk 1]
```diff
@@ -39,6 +43,8 @@ func (c *Config) Client() (interface{}, error) {
 		return nil, fmt.Errorf("Error configuring provider: when using authentication, \"key_name\", \"key_secret\" and \"key_algorithm\" should be non empty")
 	}
 	client.c = new(dns.Client)
+	client.c.Net = c.protocol
+	client.c.Timeout = time.Duration(c.timeout) * time.Second
```
Comment (katbyte) [inline dns/config.go]: What will happen when 0 is passed into the client here?

File: dns/config.go [hunk 2]
```diff
@@ -14,6 +16,8 @@ type Config struct {
 	keyname   string
 	keyalgo   string
 	keysecret string
+	protocol  string
+	timeout   int
```
Comment (katbyte) [inline dns/config.go]: You could consider using `time.ParseDuration()` here to give a friendly input such as `30s`. At the very least change it to `timeout_seconds` so the units are clear.

File: dns/config.go (hunk 1 above)
Comment (jnahelou) [reply dns/config.go]: Hello @katbyte,
Following source code : <https://github.com/miekg/dns/blob/master/client.go>
```
        // Timeout is a cumulative timeout for dial, write and read, defaults to 0 (disabled) - overrides DialTimeout, ReadTimeout,
	// WriteTimeout when non-zero. Can be overridden with net.Dialer.Timeout (see Client.ExchangeWithDialer and
	// Client.Dialer) or context.Context.Deadline (see the deprecated ExchangeContext)
	Timeout        time.Duration
	DialTimeout    time.Duration     // net.DialTimeout, defaults to 2 seconds, or net.Dialer.Timeout if expiring earlier - overridden by Timeout when that value is non-zero
	ReadTimeout    time.Duration     // net.Conn.SetReadTimeout value for connections, defaults to 2 seconds - overridden by Timeout when that value is non-zero
	WriteTimeout   time.Duration     // net.Conn.SetWriteTimeout value for connections, defaults to 2 seconds - overridden by Timeout when that value is non-zero
```

Comment (katbyte) [reply dns/config.go]: Awesome, thanks!

**Comment by katbyte** [comment]
Hi @jnahelou, I am closing this as it has been superseded by #68.

If you feel like this was in error please let us know 🙂

**Pull request closed without merging**